src/matrixrtc/NewMembershipManager.ts

```diff
@@ -0,0 +1,808 @@
+}
+
+/**
+ * @internal
+ */
+class ActionScheduler {
+    public state: ActionSchedulerState;
+
+    public constructor(
+        state: ActionSchedulerState,
+        private manager: Pick<MembershipManager, "delayedLeaveLoopHandler" | "membershipLoopHandler">,
+    ) {
+        this.state = state;
+    }
+    // state variables for a wakeup mechanism (in case we add some action externally and need to leave the current sleep)
+    private wakeupPromise?: Promise<void>;
+    private wakeup?: (value: void | PromiseLike<void>) => void;
+    private didWakeUp = false;
+
+    private actions: Action[] = [];
+    private insertions: Action[] = [];
+    private resetWith?: Action[];
+    /**
+     * This starts the main loop of the memberhsip manager that handles event sending, delayed event sending and delayed event restarting.
+     * @param initialActions The initial actions the manager will start with. It should be enough to pass: DelayedLeaveActionType.Initial
+     * @throws This throws an error only if the memberhsip cannot run anymore. For example it reached the maximum retires.
+     * In most other error cases the manager will try to handle any server errors by itself.
+     */
+    public async startWithActions(initialActions: Action[]): Promise<void> {
+        this.actions = initialActions;
+
+        while (this.actions.length > 0) {
+            this.actions.sort((a, b) => a.ts - b.ts);
+            logger.debug("Current MembershipManager action queue: ", this.actions, "\nDate.now: ", +Date.now());
+            const nextAction = this.actions[0];
+
+            this.wakeupPromise = new Promise((resolve) => {
+                this.wakeup = resolve;
+            });
+            if (nextAction.ts > Date.now()) await Promise.race([this.wakeupPromise, sleep(nextAction.ts - Date.now())]);
+            if (this.didWakeUp) {
+                // In case of a wakeup we do not want to run the next action because the next action now might be sth different.
+                // Instead we recompute the actions array and do another iteration.
+                this.didWakeUp = false;
+            } else {
+                try {
+                    if (isDelayedLeaveActionType(nextAction.type)) {
+                        await this.manager.delayedLeaveLoopHandler(this.state, nextAction.type);
+                    } else if (isMembershipActionType(nextAction.type)) {
+                        await this.manager.membershipLoopHandler(this.state, nextAction.type);
+                    }
+                } catch (e) {
+                    throw Error("The MemberhsipManager has to shut down because of the end condition: " + e);
+                }
+            }
+
+            if (this.resetWith) {
+                this.actions = this.resetWith;
+                this.resetWith = undefined;
+            }
+
+            this.actions = this.actions.filter((a) => a !== nextAction);
```

Comment:
For some reason I had expected that you would I had expected that you would have popped the original action out of the array and then you wouldn't have to filter here.  🤷‍♂️

Reply:
TLDR: I require the following order: get first element, read this element from the array where i cannot be removed in anther codeblock, then it needs to be removed.
pop makes this atomic but it is required to be two steps.

There are of course workarounds like:
```
this.currentItem = this.array.pop()
getArrayWithCurrent() return [...this.array, this.currentItem]
```
And then only use getArrayWithCurrent but I did not prefer any of the alternatives to what it is now.